### ISSUE TITLE
Robolectric tests stuck after upgrade to AGP 3.0.1

### ISSUE THREAD

**Issue by RicardoBelchior** [robolectric/robolectric#3721]
Hi,

After updating AGP from 2.3.3 to 3.0.1, my Robolectric tests randomly get stuck without finishing. The test itself finishes (i know because i printed some logs), but it seems to not move on to the next one. Plain unit tests run fine. I don't reproduce this problem in a local library module, it seems to only occur in the main app's module.

When executing: 

> ./gradlew testMirrorDebugUnitTest --debug 

I get following logs in a loop:

```
15:00:59.013 [DEBUG] [org.gradle.launcher.daemon.server.Daemon] DaemonExpirationPeriodicCheck running
15:00:59.014 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
15:00:59.014 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
15:00:59.014 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
15:00:59.015 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
15:00:59.015 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
15:00:59.015 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
15:01:01.859 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting OS memory status event {Total: 17179869184, Free: 4075302912}
15:01:01.859 [DEBUG] [org.gradle.launcher.daemon.server.health.LowMemoryDaemonExpirationStrategy] Received memory status update: {Total: 17179869184, Free: 4075302912}
15:01:01.859 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting JVM memory status event {Maximum: 3817865216, Committed: 1883766784}
15:01:06.856 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting OS memory status event {Total: 17179869184, Free: 4064776192}
15:01:06.856 [DEBUG] [org.gradle.launcher.daemon.server.health.LowMemoryDaemonExpirationStrategy] Received memory status update: {Total: 17179869184, Free: 4064776192}
15:01:06.856 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting JVM memory status event {Maximum: 3817865216, Committed: 1883766784}
15:01:09.013 [DEBUG] [org.gradle.launcher.daemon.server.Daemon] DaemonExpirationPeriodicCheck running
```
... etc.
Full log here (https://drive.google.com/open?id=1ZwQQ4_VcToHUloaFWjK-SOCzuYGnjwVq)

Things I've tried with no success:
- Adding "includeAndroidResources = true" into android > testOptions > unitTests 
- Adding to gradle.properties: 
org.gradle.daemon=false
org.gradle.parallel=false
org.gradle.configureondemand=false
android.enableAapt2=false
- Disabling test coverage

Do you guys have any clue what this might be? 

Gradle version: 4.1
Android Plugin Version: 3.0.1
Module Compile Sdk Version: 26
Module Build Tools Version: N/A
Android SDK Tools version: 27.0.3

Android Studio 3.0.1
Build #AI-171.4443003, built on November 9, 2017
JRE: 1.8.0_152-release-915-b08 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Mac OS X 10.12.6

PS: Sorry if this is a Gradle issue and I'm posting in the wrong place

**Comment by edenman** [comment]
I'm seeing this as well on some CircleCI builds.  Tried tweaking my gradle.properties Xmx settings, but that didn't fix it.

AGP 3.2.0-alpha04
Kotlin 1.2.21
Robolectric 3.7.1
Grade 4.5

Android Studio 3.2 Canary 4
Build #AI-173.4615518, built on February 21, 2018
JRE: 1.8.0_152-release-1024-b01 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Mac OS X 10.13.3

@RicardoBelchior are you still seeing this? What's the command you're running on CI? Mine is:
```
./gradlew testDebugUnitTest --profile --no-daemon -Pkotlin.incremental=false --max-workers=2 --debug --stacktrace
```
and the logs eventually just loop this segment of logs forever:
```
2:06:37.896 [DEBUG] [org.gradle.launcher.daemon.server.Daemon] DaemonExpirationPeriodicCheck running
22:06:37.896 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
22:06:37.896 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
22:06:37.896 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
22:06:37.896 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
22:06:37.896 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
22:06:37.897 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
22:06:40.848 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting OS memory status event {Total: 63314890752, Free: 47036395520}
22:06:40.848 [DEBUG] [org.gradle.launcher.daemon.server.health.LowMemoryDaemonExpirationStrategy] Received memory status update: {Total: 63314890752, Free: 47036395520}
22:06:40.848 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting JVM memory status event {Maximum: 1032847360, Committed: 1015021568}
22:06:45.848 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting OS memory status event {Total: 63314890752, Free: 47069474816}
22:06:45.848 [DEBUG] [org.gradle.launcher.daemon.server.health.LowMemoryDaemonExpirationStrategy] Received memory status update: {Total: 63314890752, Free: 47069474816}
22:06:45.848 [DEBUG] [org.gradle.process.internal.health.memory.MemoryManager] Emitting JVM memory status event {Maximum: 1032847360, Committed: 1015021568}
```
This is sounding more and more like a Gradle daemon issue, but I don't know enough about what's going on to say for sure.

**Comment by RicardoBelchior** [comment]
@edenman my case was fixed once i updated Roboletric from 3.6.1 to 3.7.1. Given the fact you're already using 3.7.1, then it was probably something else that fixed my problem. We did update a few other libraries at the same time, though.

I use the same command on CI that I use locally: _./gradlew testMirrorDebugUnitTest --debug_

I posted this same issue on Google's Issue Tracker (https://issuetracker.google.com/issues/72690634) where they suggested:

> Hi Ricardo, can you please see if your issue reproduces with the latest 3.1 Beta 4 plugin? There was a classpath issue in 3.0.1 that causes Robolectric not to work but I am not sure if it's the same issue.

Perhaps it helps your case..

**Comment by LMojica** [comment]
Have you done a `jstack` dump when your tests get stuck? I'm having the same problem, and originally I thought it was a gradle problem. After doing a `jstack` dump, it turns out that it was a deadlock within the Robolectric `Scheduler` class. I'm using AGP 3.0.0 but it also happens on 3.0.1.  I filed this issue in case you want to take a look https://github.com/robolectric/robolectric/issues/3819.

**Comment by edenman** [comment]
My jstack dump looks totally different, fwiw:
[td.txt](https://github.com/robolectric/robolectric/files/1817495/td.txt)
AFAICT, it's getting hung up on a `CountDownLatch` in `org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:69)` which looks like it's just waiting for the test to finish: https://github.com/gradle/gradle/blob/master/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/TestWorker.java#L70

There is one a "Test worker" thread waiting on a CountdownLatch for SharedPrefs, maybe that's related?

**Comment by edenman** [comment]
Tried again with Gradle 4.6.  Same issue.  The potentially problematic thread:

```
"Test worker" #22 prio=5 os_prio=0 tid=0x00007fb5ec96b800 nid=0x2c7 waiting on condition [0x00007fb554cc0000]
   java.lang.Thread.State: WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x0000000688a29680> (a java.util.concurrent.CountDownLatch$Sync)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
        at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
        at android.app.SharedPreferencesImpl$EditorImpl.$$robo$$commit(SharedPreferencesImpl.java:537)
        at java.lang.invoke.LambdaForm$DMH/821574334.invokeSpecial_L_I(LambdaForm$DMH)
        at java.lang.invoke.LambdaForm$MH/1292967378.guardWithCatch(LambdaForm$MH)
        at java.lang.invoke.LambdaForm$MH/1587898701.delegate(LambdaForm$MH)
        at java.lang.invoke.LambdaForm$MH/665095619.guard(LambdaForm$MH)
        at java.lang.invoke.LambdaForm$MH/1227675784.linkToCallSite(LambdaForm$MH)
        at android.app.SharedPreferencesImpl$EditorImpl.commit(SharedPreferencesImpl.java)
[my code]
```

**Comment by edenman** [comment]
Still an issue with Robolectric 3.8 😭

**Comment by chrispix99** [comment]
Having the same issue, saving a shared preference using commit vs. apply. Gets stuck 

"main"@1 in group "main": WAIT
park:-1, Unsafe {sun.misc}
park:175, LockSupport {java.util.concurrent.locks}
parkAndCheckInterrupt:836, AbstractQueuedSynchronizer {java.util.concurrent.locks}
doAcquireSharedInterruptibly:997, AbstractQueuedSynchronizer {java.util.concurrent.locks}
acquireSharedInterruptibly:1304, AbstractQueuedSynchronizer {java.util.concurrent.locks}
await:231, CountDownLatch {java.util.concurrent}
$$robo$$commit:537, SharedPreferencesImpl$EditorImpl {android.app}
invokeSpecial_L_I:-1, 777379084 {java.lang.invoke.LambdaForm$DMH}
guardWithCatch:-1, 1529303412 {java.lang.invoke.LambdaForm$MH}
dontInline:-1, 1641931089 {java.lang.invoke.LambdaForm$reinvoker}
guard:-1, 187019413 {java.lang.invoke.LambdaForm$MH}
linkToCallSite:-1, 1555055322 {java.lang.invoke.LambdaForm$MH}
commit:-1, SharedPreferencesImpl$EditorImpl {android.app}
updateLastSyncTime:451, SyncDependencies {com.climate.android.common.sync}
syncFinished:233, ClimateAbstractThreadedSyncAdapter {com.climate.android.common.sync}
onPerformSync:134, ClimateAbstractThreadedSyncAdapter {com.climate.android.common.sync}
runSyncAdapters:100, ScoutingSyncTest {com.climate.android.scoutinglib.sync}
testScoutingSync:75, ScoutingSyncTest {com.climate.android.scoutinglib.sync}
invoke0:-1, NativeMethodAccessorImpl {sun.reflect}
invoke:62, NativeMethodAccessorImpl {sun.reflect}
invoke:43, DelegatingMethodAccessorImpl {sun.reflect}
invoke:498, Method {java.lang.reflect}
runReflectiveCall:50, FrameworkMethod$1 {org.junit.runners.model}
run:12, ReflectiveCallable {org.junit.internal.runners.model}
invokeExplosively:47, FrameworkMethod {org.junit.runners.model}
evaluate:17, InvokeMethod {org.junit.internal.runners.statements}
evaluate:570, RobolectricTestRunner$HelperTestRunner$1 {org.robolectric}
evaluate:26, RunBefores {org.junit.internal.runners.statements}
evaluate:27, RunAfters {org.junit.internal.runners.statements}
evaluate:22, ClimateNetworkRule$1 {com.climate.android.rules}
evaluate:41, MockWebServerRule$1 {com.climate.android.rules}
run:52, PowerMockStatement$1 {org.powermock.modules.junit4.rule}
invoke0:-1, NativeMethodAccessorImpl {sun.reflect}
invoke:62, NativeMethodAccessorImpl {sun.reflect}
invoke:43, DelegatingMethodAccessorImpl {sun.reflect}
invoke:498, Method {java.lang.reflect}
performMethodInvocation:1873, WhiteboxImpl {org.powermock.reflect.internal}
doInvokeMethod:773, WhiteboxImpl {org.powermock.reflect.internal}
invokeMethod:638, WhiteboxImpl {org.powermock.reflect.internal}
invokeMethod:401, Whitebox {org.powermock.reflect}
execute:98, ClassloaderExecutor {org.powermock.classloading}
execute:78, ClassloaderExecutor {org.powermock.classloading}
evaluate:49, PowerMockStatement {org.powermock.modules.junit4.rule}
evaluate:253, SandboxTestRunner$2 {org.robolectric.internal}
runChild:130, SandboxTestRunner {org.robolectric.internal}
runChild:42, SandboxTestRunner {org.robolectric.internal}
run:290, ParentRunner$3 {org.junit.runners}
schedule:71, ParentRunner$1 {org.junit.runners}
runChildren:288, ParentRunner {org.junit.runners}
access$000:58, ParentRunner {org.junit.runners}
evaluate:268, ParentRunner$2 {org.junit.runners}
evaluate:84, SandboxTestRunner$1 {org.robolectric.internal}
run:363, ParentRunner {org.junit.runners}
run:137, JUnitCore {org.junit.runner}
startRunnerWithArgs:68, JUnit4IdeaTestRunner {com.intellij.junit4}
startRunnerWithArgs:47, IdeaTestRunner$Repeater {com.intellij.rt.execution.junit}
prepareStreamsAndStart:242, JUnitStarter {com.intellij.rt.execution.junit}
main:70, JUnitStarter {com.intellij.rt.execution.junit}

**Comment by nyjalusc** [comment]
Has anyone managed to find a fix for this? Please comment.

**Comment by edenman** [comment]
The SharedPrefs thing sounds alot like #3641 which was fixed in aea90c75f9013def854c7f1f1772f6a13cd552fb (robolectric-4.0-beta-1).  I'm still seeing occasional hangs but I'm betting it's due to another issue since it's way less predictable than this one was.  Haven't been able to get a jstack dump to confirm yet, but I think this issue can be closed out.

**Comment by jihunim** [comment]
How come is this closed?
